### DISCUSSION TITLE
lvgl9 branch on esp32-2432S028R - touch stopped working

### DISCUSSION THREAD

**Discussion by glsys** [rzeldent/esp32-smartdisplay#176] [General]
Hi, i had to switch to the 9 branch for the menu widget and started by compiling the esp32-smartdisplay-demo on the new branch - everything seems to work fine except the touch input. Any Ideas where i might look?

**Comment by rzeldent** [comment]
Hi glsys,

Nice you already tested this LVGL9 branch; I just updated it a bit without testing but seems to work already a bit...
It is a development branch and was working on it last weekend but haven't even deployed it myself!

Hopefully I'll have some time to continue working on it and this issue will hopefully be resolved quickly!

**Reply by glsys** [reply]
Thank you for your extremly fast reply!!
I am aware that this is still in early development - thats why i posted in discussions, not as issue.
I thought maybe i can be of some help with testing and bugfixing. I will have another look at it when im back at my pc.

**Reply by glsys** [reply]
still not back at my dev-pc, but they seemed to have solved it by using a tick interface: https://github.com/witnessmenow/ESP32-Cheap-Yellow-Display/commit/42faded4f0684bf3ea026b85e1a7a50775eb0bdb

**Comment by rzeldent** [comment]
Made a quick change to enable the indev device. Maybe this helps.
Also in the demo project added the tick_inc. But still untested.

**Reply by glsys** [reply]
tested it, but got a boot-loop:

`file: ".pio/libdeps/esp32-2432S028R/esp32_smartdisplay/src/lvgl_touch_xpt2046_spi.c" line 43
func: lvgl_touch_init
expression: spi_bus_initialize(XPT2046_SPI_HOST, &spi_bus_config, XPT2046_SPI_DMA_CHANNEL)
E (5293) lcd_panel.io.spi: esp_lcd_new_panel_io_spi(63): invalid argument
ESP_ERROR_CHECK failed: esp_err_t 0x102 (ESP_ERR_INVALID_ARG) at 0x40088884
  #0  0x40088884 in _esp_error_check_failed at /home/runner/work/esp32-arduino-lib-builder/esp32-arduino-lib-builder/esp-idf/components/esp_system/esp_err.c:42

file: ".pio/libdeps/esp32-2432S028R/esp32_smartdisplay/src/lvgl_touch_xpt2046_spi.c" line 62
func: lvgl_touch_init
expression: esp_lcd_new_panel_io_spi((esp_lcd_spi_bus_handle_t)XPT2046_SPI_HOST, &io_spi_config, &io_handle)`

**Reply by glsys** [reply]
strange ... i removed every folder and library and cloned everything fresh - now it woks without the error - rotation is still broken 
- going to try my ui now

**Comment by rzeldent** [comment]
That's possible because there were changes in both https://github.com/rzeldent/esp32-smartdisplay and https://github.com/rzeldent/esp32-smartdisplay-demo...

**Reply by glsys** [reply]
i looked and had both changes but somhow the stuff in .pio did not rebuild properly .. had to remove the whole directory for that (as you wrote in the Main readme :-) The new menu-widget from lvgl9 works great!! only the back button is unreachable for me because of broken touchscreencalibration ... still trying to understand how to use touch_calibration_data_t :-)

**Comment by rzeldent** [comment]
Hi Klaus,

The calibration for the touchscreen is still work in progress. This is 3 point calibration but to get the parameters you need to have both known screen and matching (obtained) coordinates to calculate. You need at least 3 points. See See: https://www.maximintegrated.com/en/design/technical-documents/app-notes/5/5296.html. Plan was to incorporate a calibrate screen but  is for the future... 

Rotation should be fixed, at least on my eso32-2432S024C (used for testing)

**Reply by glsys** [reply]
I stole the example code for line drawing to show the points and got the points out of the debug-console:

```
void showCalPoints(void)
{
    /*Create an array for the points of the line*/
    static lv_point_precise_t line_points[] = {{10,10},{10,310},{230,310} };

    /*Create style*/
    static lv_style_t style_line;
    lv_style_init(&style_line);
    lv_style_set_line_width(&style_line, 3);
    lv_style_set_line_color(&style_line, lv_palette_main(LV_PALETTE_BLUE));
    lv_style_set_line_rounded(&style_line, true);

    /*Create a line and apply the new style*/
    lv_obj_t * line1;
    line1 = lv_line_create(lv_screen_active());
    lv_line_set_points(line1, line_points, 3);     /*Set the points*/
    lv_obj_add_style(line1, &style_line, 0);
    //lv_obj_center(line1);
}
```

calling it instead of ui_init():
```
//ui_init();
showCalPoints();
```
and after that used the new values like this:

```
    const lv_point_t screen_pts[3]={{10,10},{10,310},{230,310}};
    const lv_point_t touch_pts[3]={{24,21},{24,282},{220,280}};
    touch_calibration_data = smartdisplay_compute_touch_calibration(  screen_pts, touch_pts);
```


Works like a charm!

Will try the rotation now!

**Reply by glsys** [reply]
Rotation works great too ....
tuned the calibration screen a little - but only for lvgl9
```

// 320 x 240
const lv_point_t screen_pts[3] = {{20, 20}, {20, 300}, {220, 300}}; //were the crosses are shown
const lv_point_t touch_pts[3] = {{33, 32}, {33, 273}, {209, 272}}; //fill in the found coordinates

static lv_obj_t *calpos_label;
static void showPointerPos(lv_event_t *event)
{
    lv_point_t p;
    lv_indev_t *indev = lv_event_get_indev(event);
    lv_indev_type_t indev_type = lv_indev_get_type(lv_indev_active());
    if (indev_type == LV_INDEV_TYPE_POINTER || indev_type == LV_INDEV_TYPE_BUTTON)
    {
        lv_indev_get_point(indev, &p);
        printf("X:%d, Y:%d\n", p.x, p.y);
        lv_label_set_text_fmt(calpos_label, "%d,%d", p.x, p.y);
    }
}


lv_point_precise_t line_points[6][2];
void showCalPoints(void)
{
    lv_obj_t *parent = lv_screen_active();

    /*Create style*/
    static lv_style_t style_line;
    lv_style_init(&style_line);
    lv_style_set_line_width(&style_line, 3);
    lv_style_set_line_color(&style_line, lv_palette_main(LV_PALETTE_BLUE));
    lv_style_set_line_rounded(&style_line, true);

    for (int i = 0; i < 3; i++)
    {
        int r=5;
        //horizontal
        line_points[i*2][0].x=screen_pts[i].x-r;
        line_points[i*2][0].y=screen_pts[i].y;
        line_points[i*2][1].x=screen_pts[i].x+r;
        line_points[i*2][1].y=screen_pts[i].y;

        //vertical
        line_points[i*2+1][0].x=screen_pts[i].x;
        line_points[i*2+1][0].y=screen_pts[i].y-r;
        line_points[i*2+1][1].x=screen_pts[i].x;
        line_points[i*2+1][1].y=screen_pts[i].y+r;

        lv_obj_t *line;
        line = lv_line_create(parent);
        lv_line_set_points(line, line_points[i*2], 2);
        lv_obj_add_style(line, &style_line, 0);

        line = lv_line_create(parent);
        lv_line_set_points(line, line_points[i*2+1], 2);
        lv_obj_add_style(line, &style_line, 0);

    }

    calpos_label = lv_label_create(parent);
    lv_label_set_text(calpos_label, "0,0");
    lv_obj_center(calpos_label);
    lv_obj_add_event_cb(parent, showPointerPos, LV_EVENT_PRESSED, NULL);
}
```

**Reply by glsys** [reply]
For Displays with sw_rotate =1 (ST7262) ,Rotation is broken now (works in the main branch though).
And I made a boards-file (not yet 100% shure about everything, but works) for a waveshare Board, do you see any place were this would fit in?
[esp32-s3touchlcd7.json](https://github.com/user-attachments/files/16014655/esp32-s3touchlcd7.json)

**Reply by rzeldent** [reply]
https://github.com/rzeldent/esp32-smartdisplay/issues/179 Will have to look into that issue...

**Comment by rzeldent** [comment]
Hi Klaus,

Put the board esp32-s3touchlcd7 in the project. For now in the develop branch but should be available generally in main with a new release...

Thanks for taking the effort and sharing the definitions!
